service/history/replication/dlq_writer.go

```diff
@@ -0,0 +1,73 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package replication
+
+import (
+	"context"
+
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/service/history/shard"
+)
+
+type (
+	// DLQWriter is an interface that can be implemented easily by the two different queue solutions that we have.
+	// - Queue V1 implements this interface via [persistence.ExecutionManager].
+	// - Queue V2 will implement this interface via [go.temporal.io/server/service/history/queues.DLQWriter].
+	//
+	// We want this interface to make the migration referenced by [persistence.QueueV2] easier.
+	DLQWriter interface {
+		WriteTaskToDLQ(ctx context.Context, request WriteRequest) error
+	}
+	// WriteRequest is a request to write a task to the DLQ.
+	WriteRequest struct {
+		// ShardContext is an argument that we can remove once we migrate to queue V2.
+		ShardContext        shard.Context
```

Comment:
nit: it feels to me that ideally, this should be part of the DLQWriter itself, not the Request. But I understand that means DLQWriter will be a per shard component and can't be injected from the top. It will be removed in the future anyway, so not worth making the change I think.

Reply:
Exactly. I really hated putting this here, and it was a dependency at first. However, it didn't work because that meant it couldn't be injected, especially because the `shard.Context` isn't injected in replication. It's actually fetched from the `shard.Controller`, and I can't depend on that because it creates a dependency cycle. In a subsequent PR, I made this `ExecutionManager`, which is sliiightly better, but not really. It should still be part of the object, not the request. However, it's just not possible because of the dependency on shard controller, e.g. [here](https://github.com/temporalio/temporal/blob/3f9880efb8764f55303c9e110da9e3022a280b28/service/history/replication/executable_activity_state_task.go#L175), unless we do another lazy-loaded thing, but I think it's just too complicated since it's temporary anyway.